domains/certificates/Query.API/API/CertificateGenerationSignUpService/CertificateGenerationSignUpServiceImpl.cs

```diff
@@ -0,0 +1,76 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using API.CertificateGenerationSignUpService.Clients;
+using API.CertificateGenerationSignUpService.Repositories;
+using API.MasterDataService;
+using static API.CertificateGenerationSignUpService.CreateSignUpResult;
+
+namespace API.CertificateGenerationSignUpService;
+
+internal class CertificateGenerationSignUpServiceImpl : ICertificateGenerationSignUpService
+{
+    private readonly IMeteringPointsClient client;
+    private readonly ICertificateGenerationSignUpRepository repository;
+
+    public CertificateGenerationSignUpServiceImpl(IMeteringPointsClient client, ICertificateGenerationSignUpRepository repository)
+    {
+        this.client = client;
+        this.repository = repository;
+    }
+
+    public async Task<CreateSignUpResult> Create(string gsrn, string meteringPointOwner, DateTimeOffset startDate, CancellationToken cancellationToken)
+    {
+        var meteringPoints = await client.GetMeteringPoints(meteringPointOwner, cancellationToken);
+        var matchingMeteringPoint = meteringPoints?.MeteringPoints.FirstOrDefault(mp => mp.GSRN == gsrn);
+
+        if (matchingMeteringPoint == null)
```

Comment:
Could this be extracted so the create method doesnt handle the validation?

Reply:
It could, but since the return type is a union type (or as close as we can do a union type), extracting a validation method will require that it returns a nullable result, which should then be checked in `Create`. So I prefer keeping it like this at this point.